File: state/cpp/traffic_light.cpp
#include <chrono>
#include <iostream>
#include <thread>

#include "traffic_light.h"

// Concrete states to implement requirements
 void Red::onEnter(TrafficLight& traffic_light) {
    std::cout << "Entering the Red state." << std::endl;
    std::this_thread::sleep_for(std::chrono::seconds(kRedTimeoutSec));
    std::cout << "Red light timer expired." << std::endl;
    traffic_light.set_next_state(&traffic_light.green_state);
}

void Yellow::onEnter(TrafficLight& traffic_light) {
    std::cout << "Entering the Yellow state." << std::endl;
    std::this_thread::sleep_for(std::chrono::seconds(kYellowTimeoutSec));
    std::cout << "Yellow light timer expired." << std::endl;
    traffic_light.set_next_state(&traffic_light.red_state);
}

void Green::onEnter(TrafficLight& traffic_light) {
    std::cout << "Entering the Green state." << std::endl;
    std::this_thread::sleep_for(std::chrono::seconds(kGreenTimeoutSec));
    std::cout << "Green light timer expired." << std::endl;
    traffic_light.set_next_state(&traffic_light.yellow_state);
}

TrafficLight::TrafficLight() {
    current_state = &red_state;
}

void TrafficLight::set_next_state(AbstractTrafficState* state) {
        current_state = state;
}

void TrafficLight::run(int duration_s) {
    auto start_time = std::chrono::high_resolution_clock::now();
    while (std::chrono::duration_cast<std::chrono::seconds>(std::chrono::high_resolution_clock::now() - start_time).count() < duration_s) {
        current_state->onEnter(*this);
    }
}

int main() {
    TrafficLight traffic_light;

    traffic_light.run(kSystemRunTimeSec);

    return 0;
};

File: state/python/traffic_light.py
"""
Example implementation of a traffic light system using the state design pattern
"""

from abc import ABC, abstractmethod
import time

RED_TIMEOUT_S: float = 4.0
YELLOW_TIMEOUT_S: float = 1.0
GREEN_TIMEOUT_S: float = 3.0
SYSTEM_RUN_TIME_S: float = 17.0


class InvalidStateError(Exception):
    """Error for requests to invalid states"""


# State interface that all states must adhere to
class AbstractTrafficState(ABC):
    """Interface for all traffic light states to implement"""

    @abstractmethod
    def enter(self, traffic_light: "TrafficLight") -> None:
        """Behavior when entering the state.

        :param traffic_light: TrafficLight system (state handler)
        """


# Concrete states to meet requirements
class Red(AbstractTrafficState):
    """Implementation of the behaviors within the Red light state."""

    def enter(self, traffic_light: "TrafficLight") -> None:
        print("Entering the Red state.")
        time.sleep(RED_TIMEOUT_S)
        print("Red light timer expired.")
        traffic_light.next_state = Green()


class Yellow(AbstractTrafficState):
    """Implementation of the behaviors within the Yellow light state."""

    def enter(self, traffic_light: "TrafficLight") -> None:
        print("Entering the Yellow state.")
        time.sleep(YELLOW_TIMEOUT_S)
        print("Yellow light timer expired.")
        traffic_light.next_state = Red()


class Green(AbstractTrafficState):
    """Implementation of the behaviors within the Green light state"""

    def enter(self, traffic_light: "TrafficLight") -> None:
        print("Entering the Green state.")
        time.sleep(GREEN_TIMEOUT_S)
        print("Green light timer expired.")
        traffic_light.next_state = Yellow()


class NullState(AbstractTrafficState):
    """Empty state for initialization"""

    def enter(self, traffic_light: "TrafficLight") -> None:
        raise InvalidStateError("Invalid state requested!!")


# State machine handler - representation of state machine
class TrafficLight:
    """Implementation of the traffic light state machine"""

    def __init__(self) -> None:
        self.red: AbstractTrafficState = Red()
        self.yellow: AbstractTrafficState = Yellow()
        self.green: AbstractTrafficState = Green()
        self.null: AbstractTrafficState = NullState()

        self.current_state: AbstractTrafficState = self.red
        self.next_state: AbstractTrafficState = self.null

    def run(self, duration_s: float = SYSTEM_RUN_TIME_S) -> None:
        """Runs the traffic light state machine"""
        start_time: float = time.time()
        while time.time() - start_time < duration_s:
            self.current_state.enter(self)
            self.current_state = self.next_state


def main() -> None:
    """Script for example state machine execution"""
    traffic_light = TrafficLight()

    # Run the state machine
    traffic_light.run()


if __name__ == "__main__":
    main()
